### PR TITLE
add build for complex scalars

### PR DESCRIPTION
**Pull request by ltalirz**
Add build of library for complex scalars.

This is an attempt at addressing issue #17 
(disclaimer: I'm not very familiar with petsc; I just need the complex variant for another package and I've read through the issue thread)

Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by ltalirz** [comment]
@conda-forge-admin, please rerender

**Comment by dalcinl** [comment]
How are users supposed to install the real build vs the complex build?

**Comment by ltalirz** [comment]
> How are users supposed to install the real build vs the complex build?

If I'm not mistaken, you should be able to install it by selecting the variant. E.g.
```
conda install -c conda-forge petsc=*=*complex*
```

**Comment by dalcinl** [comment]
@isuruf I need your input here. This PR is adding support for `real` and `complex` build variants of PETSc. The default should not change, and it should be `real`. What's the proper way to configure things?

**Comment by ltalirz** [comment]
@dalcinl Coming back to this: I think the previous CI failure was spurious - a simple re-push made them go away.

Ready for review!

**Comment by dalcinl** [comment]
@isuruf Why are you sure the `real_scalar` build is still the default? Because of `track_features`?
Do we really need the longish `{real|complex}_scalar` name in the build string and `track_features`? Why not just `{real|complex}`, which IMHO is descriptive enough about what it is about?

**Comment by ltalirz** [comment]
> @isuruf Why are you sure the real_scalar build is still the default? Because of track_features?

Indeed, you can use it to de-prioritize specific variants, see [here](https://docs.conda.io/projects/conda/en/latest/user-guide/concepts/packages.html#track-features).

I guess including the name of the variable in the build string is in principle safer in case it ever also becomes part of another variant, but it may be ok to drop in this case. Let's see what @isuruf has to say

**Comment by isuruf** [comment]
Yes, `track_features` de-prioritizes complex variant and I've also added higher build number for real, which prioritizes real variant.

I don't have an opinion on `real_scalar` vs `real` in the build string.

File: recipe/meta.yaml [hunk 1]
```diff
@@ -2,6 +2,10 @@
 {% set sha256 = "ac46db6bfcaaec8cd28335231076815bd5438f401a4a05e33736b4f9ff12e59a" %}
 
 {% set mpi = mpi or 'mpich' %}
+{% set build = 1 %}
```
Comment (dalcinl) [inline recipe/meta.yaml:5]: @ltalirz Could you please move this line setting the build number right after the sha256 line above ?

**Comment by ltalirz** [comment]
thanks for the review @dalcinl , done!

**Comment by ltalirz** [comment]
I'm going to prepare an analogous PR for slepc

**Comment by TLKaiser** [comment]
Thanks for all your work! I am, too, highly interested in getting PETSc and SLEPc to work for complex numbers. I am having a little problem though with the uploaded build. When I install the 'real' build of PETSc then 

`from petsc4py import PETSc`
runs perfectly fine. When I install the 'complex' build, then I get the following error

> >>> from petsc4py import PETSc
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
>   File "/Users/tkaiser/opt/anaconda3/envs/test3/lib/python3.9/site-packages/petsc4py/PETSc.py", line 3, in <module>
>     PETSc = ImportPETSc(ARCH)
>   File "/Users/tkaiser/opt/anaconda3/envs/test3/lib/python3.9/site-packages/petsc4py/lib/__init__.py", line 29, in ImportPETSc
>    return Import('petsc4py', 'PETSc', path, arch)
>   File "/Users/tkaiser/opt/anaconda3/envs/test3/lib/python3.9/site-packages/petsc4py/lib/__init__.py", line 73, in Import
>     module = import_module(pkg, name, path, arch)
>   File "/Users/tkaiser/opt/anaconda3/envs/test3/lib/python3.9/site-packages/petsc4py/lib/__init__.py", line 58, in import_module
>     with f: return imp.load_module(fullname, f, fn, info)
>   File "/Users/tkaiser/opt/anaconda3/envs/test3/lib/python3.9/imp.py", line 242, in load_module
>     return load_dynamic(name, filename, file)
>   File "/Users/tkaiser/opt/anaconda3/envs/test3/lib/python3.9/imp.py", line 342, in load_dynamic
>     return _load(spec)
> ImportError: dlopen(/Users/tkaiser/opt/anaconda3/envs/test3/lib/python3.9/site-packages/petsc4py/lib/PETSc.cpython-39-darwin.so, 2): Symbol not found: _TaoBRGNGetDampingVector
>   Referenced from: /Users/tkaiser/opt/anaconda3/envs/test3/lib/python3.9/site-packages/petsc4py/lib/PETSc.cpython-39-darwin.so
>   Expected in: /Users/tkaiser/opt/anaconda3/envs/test3/lib/libpetsc.3.15.dylib
>  in /Users/tkaiser/opt/anaconda3/envs/test3/lib/python3.9/site-packages/petsc4py/lib/PETSc.cpython-39-darwin.so`


Am I missing a step, or is there a problem with the build? I get the same error on a CentOS-7 machine.

**Comment by ltalirz** [comment]
Thanks for the report @TLKaiser - I haven't yet gotten round to actually incorporating the complex build into the project I wanted to so it's quite possible that there are still things to fix.

However, I'm not a petsc expert and it is not clear to me whether the problem you notice (the missing ` _TaoBRGNGetDampingVector` symbol) is a problem of the petsc library missing the symbol or of the `petsc4py` wrapper expecting the symbol to be there.
Maybe @dalcinl can enlighten us?

**Comment by dalcinl** [comment]
@ltalirz you implemented complex support for PETSc and SLEPc. But there are no matching petsc4py/slepc4py builds. I guess that @TLKaiser 's  problem is that the petsc4py package conda is installing is one built for real scalars, but the PETSc package is a complex build (and then the shared library is missing the TAO solvers, which are not available for complex builds).

**Comment by ltalirz** [comment]
@dalcinl are you saying that one may just need a build of petsc4py that depends on the complex variant of the petsc library instead of the real one or is there more to it?
Is petsc4py ready for complex input?

I currently have little time to work on it; of someone else wants to give it a try, please go ahead

**Comment by dalcinl** [comment]
@ltalirz Yes, that's exactly what I'm saying. petsc4py has been always ready for complex scalars. Updating the recipe to add complex builds is probably trivial, however I'm not sure what's the proper approach.

**Comment by ltalirz** [comment]
PR in https://github.com/conda-forge/petsc4py-feedstock/pull/40

**Pull request merged**